webapp/src/app.tsx

```diff
@@ -76,7 +76,11 @@ let theEditor: ProjectView;
 let hash: { cmd: string, arg: string };
 let pendingEditorRequests: ((p: ProjectView) => void)[];
 
-function getEditorAsync() {
+export function hasEditor() {
+    return !!theEditor;
+}
+
+export function getEditorAsync() {
```

Comment:
I'm really unsure about this approach. The goal is to detect if we have an editor open and if so allow getting a handle so we can change which project is being edited.

However, these weren't public before and it doesn't seem great to allow arbitrary parts of the code to reach and mutate the editor like this.

Yet I don't know of a better way with our current abstractions.

(This is another case where something like redux would be great: I'd like to change the currently edited project state and that should cascade automatically into updating the relevant editors if needed)

Reply:
Agreed this isn't great. Virtual api is our redux analog. Can it be used here to load the new header?

Reply:
The problem is it doesn't support a quarter of the things proper state management in React or Redux would do. For example, each header load causes at least one save to happen as a biproduct which invalidates the header. If the editor reacted to header invalidations, we'd be in an endless loop. The virtual API invalidation is so noisy around headers that it really can't be relied upon for many tasks.